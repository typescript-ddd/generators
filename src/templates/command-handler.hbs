import { CommandHandler } from "@typescript-ddd/core";
import { {{actorName}} } from "../../application";
import { {{commandName}} } from "./{{commandName}}";

/**
 * Represents a handler for the {{natural actionName}} {{singular (natural entityName)}} commmand.
 */
export class {{commandName}}Handler implements CommandHandler<{{commandName}}> {
  /**
   * Initializes a new {{natural actionName}} {{natural entityName}} command handler instance.
   * @param {{lbrace}}{{pascal entityName}}{{actionActor actionName}}{{rbrace}} {{natural (actionActor actionName)}} - the {{natural entityName}} {{natural (actionActor actionName)}}.
   */
  constructor(private readonly {{lower (actionActor actionName)}}: {{actorName}}) {}

  /**
   * Handles the {{natural actionName}} {{natural entityName}} command.
   * @param {{lbrace}}{{commandName}}{{rbrace}} command - The command to handle.
   * @returns {Promise<void>}
   */
  public async handle(command: {{commandName}}): Promise<void> {
    await this.{{lower (actionActor actionName)}}.{{actionName}}({
      {{#each properties}}
      {{name}}: command.{{name}},
      {{/each}}
    });
  }
}
