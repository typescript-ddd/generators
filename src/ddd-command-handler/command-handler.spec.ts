import { generateCommandHandler } from "./command-handler.generator";
import { CommandHandlerInput } from "./command-handler.types";

describe("CommandHandler", () => {
  it("should render with create action", () => {
    // Arrange
    const data: CommandHandlerInput = {
      entityName: "User",
      actionName: "create",
      properties: [
        { name: "name", type: "string" },
        { name: "email", type: "string" },
      ],
    };

    // Act
    const result = generateCommandHandler(data);

    // Assert
    expect(result).toMatchSnapshot();
  });

  it("Should render with update action", () => {
    // Arrange
    const data: CommandHandlerInput = {
      entityName: "User",
      actionName: "update",
      properties: [
        { name: "name", type: "string" },
        { name: "email", type: "string" },
      ],
    };
    // Act
    const result = generateCommandHandler(data);
    // Assert
    expect(result).toMatchSnapshot();
  });

  it("Should render with delete action", () => {
    // Arrange
    const data: CommandHandlerInput = {
      entityName: "User",
      actionName: "delete",
      properties: [],
    };
    // Act
    const result = generateCommandHandler(data);
    // Assert
    expect(result).toMatchSnapshot();
  });
});
