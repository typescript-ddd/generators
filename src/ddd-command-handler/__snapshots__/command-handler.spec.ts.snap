// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandHandler Should render with delete action 1`] = `
"import { CommandHandler } from "@typescript-ddd/core";
import { UserDeleter } from "../../application";
import { DeleteUserCommand } from "./DeleteUserCommand";

/**
 * Represents a handler for the delete user commmand.
 */
export class DeleteUserCommandHandler implements CommandHandler<DeleteUserCommand> {
  /**
   * Initializes a new delete user command handler instance.
   * @param {UserDeleter} deleter - the user deleter.
   */
  constructor(private readonly deleter: UserDeleter) {}

  /**
   * Handles the delete user command.
   * @param {DeleteUserCommand} command - The command to handle.
   * @returns {Promise<void>}
   */
  public async handle(command: DeleteUserCommand): Promise<void> {
    await this.deleter.delete({
      id: command.id,
    });
  }
}
"
`;

exports[`CommandHandler Should render with update action 1`] = `
"import { CommandHandler } from "@typescript-ddd/core";
import { UserUpdater } from "../../application";
import { UpdateUserCommand } from "./UpdateUserCommand";

/**
 * Represents a handler for the update user commmand.
 */
export class UpdateUserCommandHandler implements CommandHandler<UpdateUserCommand> {
  /**
   * Initializes a new update user command handler instance.
   * @param {UserUpdater} updater - the user updater.
   */
  constructor(private readonly updater: UserUpdater) {}

  /**
   * Handles the update user command.
   * @param {UpdateUserCommand} command - The command to handle.
   * @returns {Promise<void>}
   */
  public async handle(command: UpdateUserCommand): Promise<void> {
    await this.updater.update({
      id: command.id,
      name: command.name,
      email: command.email,
    });
  }
}
"
`;

exports[`CommandHandler should render with create action 1`] = `
"import { CommandHandler } from "@typescript-ddd/core";
import { UserCreator } from "../../application";
import { CreateUserCommand } from "./CreateUserCommand";

/**
 * Represents a handler for the create user commmand.
 */
export class CreateUserCommandHandler implements CommandHandler<CreateUserCommand> {
  /**
   * Initializes a new create user command handler instance.
   * @param {UserCreator} creator - the user creator.
   */
  constructor(private readonly creator: UserCreator) {}

  /**
   * Handles the create user command.
   * @param {CreateUserCommand} command - The command to handle.
   * @returns {Promise<void>}
   */
  public async handle(command: CreateUserCommand): Promise<void> {
    await this.creator.create({
      name: command.name,
      email: command.email,
    });
  }
}
"
`;
